### PR TITLE
Fast Forwarding Previous Struct Work

### PR DESCRIPTION
**Pull request by skippi**
This PR focuses on fixing struct-specific features that should have been or should now be changed due to decisions we've made in the most recent PRs. Many of these include corrections to fix inconsistencies. Pretty much a fast-forward PR. This includes:

* Adding mentions to specific structs.
* Improving documentation to match our current standard.
* Improving typespecs where possible.
* Improving function names where possible.
* Creating and improving tests and doctests.

I'll hit API specific features in a future PR.

### CODE REVIEW

**Comment by skippi** [comment]
I think this PR is done, but let me know if it needs to be broken up. It is a pretty big PR that covers a lot of different changes.

File: test/nostrum/struct/user_test.exs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+defmodule Nostrum.Struct.UserTest do
+  use ExUnit.Case, async: true
+
+  alias Nostrum.Struct.User
+
+  doctest User
+
+  describe "String.Chars" do
```
Comment (Kraigie) [inline test/nostrum/struct/user_test.exs]: I think it's a little redundant to have a doctest and an actual test. What do you think? This is the case throughout this pr. Maybe it doesn't matter?

Comment (skippi) [reply test/nostrum/struct/user_test.exs]: They aren't redundant, but I do wish I could have made them doctests. They test functionality for the `String.Chars` protocol, but the protocols aren't directly doctestable IIRC. If there is a way to make them doctestable, I would be on board.

**Comment by Kraigie** [comment]
Thanks!

**Pull request merged**